{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1eaa5729",
   "metadata": {},
   "source": [
    "## Graph\n",
    "\n",
    "**Graph** is a pair of sets `(V,E)`, where `V` - array of vertices(nodes) and `E` - array of edges. For example `V = {a,b,c,d,e,f,g,h,i}` and E = `{ (a,b); (b,c); (c,e); (e,h); (h,i); (c,i)}`\n",
    "\n",
    "**A directed graph** or digraph is a graph in which edges have orientations.\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/2/23/Directed_graph_no_background.svg/1280px-Directed_graph_no_background.svg.png)\n",
    "\n",
    "**A weighted graph** is a graph in which a number (the weight) is assigned to each edge.\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/f/f0/Weighted_network.svg/1920px-Weighted_network.svg.png)\n",
    "\n",
    "**A multigraph** is a graph which is permitted to have multiple edges (also called parallel edges[1]), that is, edges that have the same end nodes. Thus two vertices may be connected by more than one edge.\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Multi-pseudograph.svg/800px-Multi-pseudograph.svg.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f3de9cc",
   "metadata": {},
   "source": [
    "## Connectivity \n",
    "\n",
    "**Path** from $v_0$ to $v_4$: {$v_0e_1v_1e_2v_2e_3v_3e_4v_4$}\n",
    "\n",
    "In an undirected graph G, two vertices u and v are called **connected** if G contains a path from u to v.\n",
    "\n",
    "A graph is said to be **connected** if every pair of vertices in the graph is connected.\n",
    "\n",
    "A directed graph is called **weakly connected** if replacing all of its directed edges with undirected edges produces a connected (undirected) graph. \n",
    "\n",
    "It is **strongly connected**, or simply strong, if it contains a directed path from u to v and a directed path from v to u for every pair of vertices u, v.\n",
    "\n",
    "A **connected component** is a maximal connected subgraph of an undirected graph. Each vertex belongs to exactly one connected component\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/e/e1/Scc-1.svg/440px-Scc-1.svg.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00f5230d",
   "metadata": {},
   "source": [
    "# DFS\n",
    "\n",
    "![alternatvie text](https://www.codesdope.com/staticroot/images/algorithm/dfs.gif)\n",
    "\n",
    "Time complexity: $O(V+E)$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "693b4486",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 0 1 3 "
     ]
    }
   ],
   "source": [
    "class Graph:\n",
    "    def __init__(self):\n",
    "         self.graph = {}\n",
    " \n",
    "    # function to add an edge to graph\n",
    "    def addEdge(self, u, v):\n",
    "        if u not in self.graph:\n",
    "            self.graph[u] = []\n",
    "        self.graph[u].append(v)\n",
    " \n",
    "    def dfs(self, v, visited):\n",
    "        visited.add(v)\n",
    "        print(v, end=' ')\n",
    "        for neighbour in self.graph[v]:\n",
    "            if neighbour not in visited:\n",
    "                self.dfs(neighbour, visited)\n",
    " \n",
    "    def DFSMain(self, v):\n",
    "        visited = set()\n",
    "        self.dfs(v, visited)\n",
    "        \n",
    "g = Graph()\n",
    "g.addEdge(0, 1)\n",
    "g.addEdge(0, 2)\n",
    "g.addEdge(1, 2)\n",
    "g.addEdge(2, 0)\n",
    "g.addEdge(2, 3)\n",
    "g.addEdge(3, 3)\n",
    "g.DFSMain(2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd11a0e3",
   "metadata": {},
   "source": [
    "\n",
    "The result of a depth-first search of a graph can be conveniently described in terms of a **spanning tree** of the vertices reached during the search:\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/5/57/Tree_edges.svg/1280px-Tree_edges.svg.png)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "72d01855",
   "metadata": {},
   "outputs": [],
   "source": [
    "entry = {}\n",
    "leave = {} # dicts of entry and leave time points for each node\n",
    "def dfs(u, time):\n",
    "    entry[u] = time+1\n",
    "    #...\n",
    "    leave[u] = time+1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb4aa90",
   "metadata": {},
   "source": [
    "### Lemma 1\n",
    "\n",
    "During depth-first search there is no time moment where there is an edge from black node to white in DFS path.\n",
    "\n",
    "**Proof**\n",
    "\n",
    "Assume that such an edge (u,v) exists at moment *time*. leave[u] is the first moment when  u is black and \n",
    "\n",
    "$leave[u] \\leq time$\n",
    "\n",
    "That means that v is white at leave[u] as it is white at moment *time*. However, this means that at the moment of leaving the vertex u, we have an unprocessed edge (u,v) left. Contradiction.\n",
    "\n",
    "### Lemma 2\n",
    "\n",
    "Assume some dfs path exists in the graph G. entry[u] and leave[u] are entry and leave time points for node u. Then between these moments:\n",
    "\n",
    "- Black and gray vertices *G\\u* will not change their color\n",
    "\n",
    "- White vertices *G\\u* will either stay white or turn black. Moreover, those who are reachable from u along white paths and only they will become black.\n",
    "\n",
    "Proof\n",
    "\n",
    "Black vertex stay black.\n",
    "\n",
    "The gray vertex will remain gray because it is on the recursion stack.\n",
    "\n",
    "The reachable white vertex will become black, otherwise on the way to it at time moment leave[u] there will be an edge from the black vertex to the white one\n",
    "\n",
    "If the vertex turned black, then it was reachable by the white path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e675425",
   "metadata": {},
   "source": [
    "- We can find cycles using DFS\n",
    "\n",
    "If we found a gray vertex during dfs (back edge), then the cycle exists.\n",
    "\n",
    "If the cycle exists, then we will definitely find it. Let v be the first vertex of cycle C processed by dfs. At time entry[v], all vertices of cycle C are white. There is an edge (u,v) of the cycle C. By Lemma 2, the vertex u becomes black before leaving v. Therefore, at the moment entry[u], the vertex in will be gray.\n",
    "\n",
    "\n",
    "- We can count connected components in undirected graph during DFS:\n",
    "\n",
    "Increasing the number of connected components until we visit all vertices during dfs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581d494c",
   "metadata": {},
   "source": [
    "## Topological sorting\n",
    "\n",
    "We can sort an acyclic graph $G(V,E)$. Topological sorting is the ordering of the vertices V such that if $(u,v) \\in E$ then u is located in an ordered array before v. There can be many valid topological sorts.:\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/thumb/0/03/Directed_acyclic_graph_2.svg/610px-Directed_acyclic_graph_2.svg.png)\n",
    "\n",
    "- 5, 7, 3, 11, 8, 2, 9, 10 (visual top-to-bottom, left-to-right)\n",
    "- 3, 5, 7, 8, 11, 2, 9, 10 (smallest-numbered available vertex first)\n",
    "- 5, 7, 3, 8, 11, 10, 9, 2 (fewest edges first)\n",
    "- 7, 5, 11, 3, 10, 8, 9, 2 (largest-numbered available vertex first)\n",
    "- 5, 7, 11, 2, 3, 8, 9, 10 (attempting top-to-bottom, left-to-right)\n",
    "- 3, 7, 8, 5, 11, 10, 2, 9 (arbitrary)\n",
    "\n",
    "\n",
    "For each vertex v, we can set $\\phi(v) = |V|+1-leave[v]$ as a position in topologically sorted array.\n",
    "\n",
    " $\\phi(u) < \\phi(v)$ for each $(u,v) \\in E$:\n",
    " \n",
    "At moment entry[u] v -is not gray (acyclic graph)\n",
    "\n",
    "1) If v is white it will be processed during dfs(u) \n",
    "\n",
    "leave[v] < leave[u] => $\\phi(u) < \\phi(v)$\n",
    "\n",
    "2) If v is black - it was already processed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ee10bce",
   "metadata": {},
   "source": [
    "## Kosaraju's algorithm\n",
    "\n",
    "How to find strongly connected components in graph $G(V,E)$?\n",
    "\n",
    "1) We construct a transpose graph $H = G^t$\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/d/d0/Amirali_reverse.jpg)\n",
    "\n",
    "\n",
    "2) DFS(H), collect all $leave_H[v]$\n",
    "\n",
    "3) DFS(G), iterate over the vertices in descending order of their $leave_H[v]$. Spanning trees of this DSF call contain the vertices of strongly connected components.\n",
    "\n",
    "- First, we will prove that each strongly connected component (SCC) is completely contained in a tree:\n",
    "\n",
    "t and s are vertices from one SCC. Then the paths from s to t and from t to s exist. v is the first vertex in dfs path s -> t -> s. Then at moment entry[v] s and t are reachable from v via white paths and according to Lemma 2 they will be processed.\n",
    "\n",
    "![alternatvie text](img/Kosaraju_1.png)\n",
    "\n",
    "- Second, one tree contains only one SCC\n",
    "\n",
    "If C - SCC, then leave[C]=max(leave[v]) for all $v \\in C$ \n",
    "\n",
    "Lemma\n",
    "\n",
    "C, C'  - SCCs, edge (u,v) connects C and C'. Then leave[C] > leave[C']. \n",
    "\n",
    "a) If C was processed before C': w is first vertex in C during dfs. At the time of entering the C component, the entire C' component is white. leave[w] > leave[C']\n",
    "\n",
    "b) If C' was processed before C: No path from C' to C exists => whole C' will be processed before C and leave[C] > leave[C']\n",
    "\n",
    "![alternatvie text](img/Kosaraju_2.png)\n",
    "\n",
    "Let's show that one tree T contains only one SCC.\n",
    "\n",
    "If T contains two SCC: C and C' and C is the first component processed in DFS(G). Edge (u,v) that connects C and C' exists.\n",
    "\n",
    "$leave_H[C] > leave_H[C'] $ (This follows from how we constructed T)\n",
    "\n",
    "However $leave_H[C] < leave_H[C'] $ follows from Lemma if it is applied to DFS(H).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927ce33e",
   "metadata": {},
   "source": [
    "# Breadth first search\n",
    "\n",
    "Vertices are processed in order of increasing distance from the starting vertex. Here we use queue data structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "099bec80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 0 3 1 "
     ]
    }
   ],
   "source": [
    "from collections import deque\n",
    "class Graph:\n",
    "    def __init__(self):\n",
    "        self.graph = {}\n",
    "        self.vertices = set()\n",
    " \n",
    "    # function to add an edge to graph\n",
    "    def addEdge(self, u, v):\n",
    "        if u not in self.graph:\n",
    "            self.graph[u] = []\n",
    "        self.graph[u].append(v)\n",
    "        self.vertices.add(u)\n",
    "        self.vertices.add(v)\n",
    "\n",
    "    def bfs(self, v):\n",
    "        visited = {}\n",
    "        for u in self.vertices:\n",
    "            visited[u] = False\n",
    "        queue = deque([])\n",
    "        queue.append(v)\n",
    "        visited[v] = True\n",
    "        while queue:\n",
    "            current_node = queue.popleft()\n",
    "            print(current_node, end=' ')\n",
    "            for u in self.graph[current_node]:\n",
    "                if not visited[u]:\n",
    "                    queue.append(u)\n",
    "                    visited[u] = True\n",
    "                \n",
    "g = Graph()\n",
    "g.addEdge(0, 1)\n",
    "g.addEdge(0, 2)\n",
    "g.addEdge(1, 2)\n",
    "g.addEdge(2, 0)\n",
    "g.addEdge(2, 3)\n",
    "g.addEdge(3, 3)\n",
    "g.bfs(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf3e6a7c",
   "metadata": {},
   "source": [
    "- Black - vertex that has been extracted from the queue\n",
    "\n",
    "- Gray - vertex that is in the queue\n",
    "\n",
    "- White - vertex that has not yet been processed\n",
    "\n",
    "At each moment of time in the process of BFS, there are vertices in the queue that are at a distance k from the start, and behind them - at a distance k + 1.\n",
    "\n",
    "##  Shortest paths\n",
    "Using this algorithm, we can find the shortest paths from a given vertex to all other reachable vertices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c7f494d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0: 1, 1: 2, 2: 0, 3: 1}\n"
     ]
    }
   ],
   "source": [
    "from collections import deque\n",
    "class Graph:\n",
    "    def __init__(self):\n",
    "        self.graph = {}\n",
    "        self.vertices = set()\n",
    " \n",
    "    # function to add an edge to graph\n",
    "    def addEdge(self, u, v):\n",
    "        if u not in self.graph:\n",
    "            self.graph[u] = []\n",
    "        self.graph[u].append(v)\n",
    "        self.vertices.add(u)\n",
    "        self.vertices.add(v)\n",
    "\n",
    "    def bfs(self, v):\n",
    "        dist = {} # shortest distance to given vertex\n",
    "        pi = {} # previous node from the shortest path\n",
    "        for u in self.vertices:\n",
    "            dist[u] = float(\"inf\")\n",
    "            pi[u] = -1\n",
    "        queue = deque([])\n",
    "        queue.append(v)\n",
    "        dist[v] = 0\n",
    "        pi[v] = -1\n",
    "        while queue:\n",
    "            current_node = queue.popleft()\n",
    "            for u in self.graph[current_node]:\n",
    "                if dist[u] > dist[current_node]+1:\n",
    "                    dist[u] = dist[current_node]+1\n",
    "                    pi[u] = current_node\n",
    "                    queue.append(u)\n",
    "        print(dist)\n",
    "                \n",
    "g = Graph()\n",
    "g.addEdge(0, 1)\n",
    "g.addEdge(0, 2)\n",
    "g.addEdge(1, 2)\n",
    "g.addEdge(2, 0)\n",
    "g.addEdge(2, 3)\n",
    "g.addEdge(3, 3)\n",
    "g.bfs(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fa897b",
   "metadata": {},
   "source": [
    "#  Eulerian walk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44c5447",
   "metadata": {},
   "source": [
    "In an undirected graph, find a path (or cycle) that passes through all the edges of the graph once. The corresponding path or cycle is called Eulerian.\n",
    "\n",
    "![alternatvie text](https://networkx.org/nx-guides/_images/part1.png)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fad29cbe",
   "metadata": {},
   "source": [
    "G - connected undirected graph\n",
    "\n",
    "An Euler path exists if and only if G has at most two odd vertices\n",
    "\n",
    "An Euler cycle exists if and only if all vertices are even.\n",
    "\n",
    "\n",
    "=> \n",
    "\n",
    "When we process some vertex, we use 2 edges, all intermediate vertices included to the path should be even. \n",
    "\n",
    "<= \n",
    "\n",
    "An Euler cycle: \n",
    "- we find some cycle C during search, exclude it from G. \n",
    "- Repeat for each component of a new graph G\\C. \n",
    "- Combine all found cycles into one. \n",
    "\n",
    "An Euler path: \n",
    "\n",
    "- build a path from an odd vertex to an odd one.\n",
    "- Subtract it from the graph. \n",
    "- repeat the same procedure as for the Euler cycle\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30b942df",
   "metadata": {},
   "source": [
    "# De novo Assembly\n",
    "\n",
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/b/b6/Types_of_sequencing_assembly.png)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb03f8a",
   "metadata": {},
   "source": [
    "## Suffix-prefix match\n",
    "\n",
    "Reads that come from the same region of the genome can overlap:\n",
    "\n",
    "![alternatvie text](img/suf_pref.png)\n",
    "\n",
    "Mismatches because of:\n",
    "\n",
    "- sequencing errors\n",
    "- polyploidy\n",
    "\n",
    "More coverage - more and longer overlaps\n",
    "\n",
    "## Overlap graph\n",
    "\n",
    "Vertices are sequences, they are connected with directed edge if they overlap (from a vertex that have a common substring as a suffix to a vertex where it is a prefix). Each edge is lables with a length of overlap\n",
    "\n",
    "We set some threshold length of an overlap between nodes\n",
    "\n",
    "![alternatvie text](img/overlap_graph.png)\n",
    "\n",
    "\n",
    "\n",
    "A sequence of the original genome can be build by walking a certain path in this graph\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f8c0210",
   "metadata": {},
   "source": [
    "## Shortest common superstring\n",
    "\n",
    "Given a set of strings S, find shortest string that contains all strings in S as substrings.\n",
    "\n",
    "NP-complete\n",
    "\n",
    "This problem is similar to the builing assembly from reads\n",
    "\n",
    "Algorithm result depends on the order we concatinate strings\n",
    "\n",
    "![alternatvie text](img/scs_not_greedy.png)\n",
    "\n",
    "\n",
    "n! different orderings to check\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dd391be",
   "metadata": {},
   "source": [
    "## Greedy algorithm\n",
    "\n",
    "At each time we neew to peek an edge that represents the lognest overlap in overlap graph\n",
    "\n",
    "Result is not always optimal\n",
    "\n",
    "![alternatvie text](img/scs_greedy.png)\n",
    "\n",
    "**Such a greedy algroritm can eliminate repeats that are present in common string (genome) -> we want to update our model** \n",
    "\n",
    "##  De Bruijn graph\n",
    "\n",
    "Building a De Bruijn graph helps us to overcome this repeat collapsing problem\n",
    "\n",
    "Let's assume that our reads are k-mers from the genome\n",
    "\n",
    "Split each k-mer to left and right k-1-mer (1 base difference) - L and R, draw an edge from L to R\n",
    "\n",
    "Each k-mer in the genome corresponds to one **edge** in this graph\n",
    "\n",
    "![alternatvie text](img/de_bruijn.png)\n",
    "\n",
    "How to construct an original sequence from De Bruijn graph -> Eulerian path in graph\n",
    "\n",
    "What if there is more than one Eulerian path in graph?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e579e0f",
   "metadata": {},
   "source": [
    "## Edge-connection\n",
    "\n",
    "Vertices u and v in an *undirected graph* are **edge-connected** if there are two paths between these vertices that do not intersect in edges (edge-disjoint paths).\n",
    "\n",
    "![alternatvie text](img/edge_connection.png)\n",
    "\n",
    "\n",
    "Edge-connectivity is an equivalence relation:\n",
    "\n",
    "- $u \\sim u$\n",
    "- $u \\sim v \\implies v \\sim u$\n",
    "- $u \\sim v, v \\sim w \\implies u \\sim w$\n",
    "\n",
    "\n",
    "Let's show transitivity:\n",
    "\n",
    "1) $u \\sim v \\implies$ We define cycle $c$ - a union of two edge-disjoint paths from u to v\n",
    "\n",
    "2) $v \\sim w \\implies$  P1 and P2 are the two edge-disjoint paths from w to v. P1 and P2 intersect C in 2 vertices - a and b; a or b can be the same vertex as v\n",
    "\n",
    "We can build 2 edge-disjoint paths from w to u:\n",
    "\n",
    "- u -> a -> w\n",
    "\n",
    "- u -> v -> b -> w\n",
    "\n",
    "![alternatvie text](img/edge_conn_trans.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6860e512",
   "metadata": {},
   "source": [
    "All graph vertices are partitioned into edge-connected equivalence classes:\n",
    "\n",
    "![alternatvie text](img/bridges.png)\n",
    "\n",
    "**Bridge**\n",
    "is:\n",
    "\n",
    "1) an edge connecting 2 different edge-connected components\n",
    "\n",
    "2) an edge, upon removal of which the connected component breaks up\n",
    "\n",
    "3) an edge that lies on any path connecting u with v.\n",
    "\n",
    "$1 \\implies 2$ : If we delete bridge edge, we won't find the second path between two vertices connected by the bridge\n",
    "\n",
    "$2 \\implies 3$ : If other path from u to v that do not contain a bridge exist, then the connected component would not break up\n",
    "\n",
    "$3 \\implies 1$ : If u and v belong to the same edge-connected component,then 2 edge-disjoint paths exist between u and v\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84b1f8db",
   "metadata": {},
   "source": [
    "### Finding bridges in graph\n",
    "\n",
    "We can find bridges in graph using DFS: \n",
    "\n",
    "Suppose that during DFS, we pass an edge from vertex v to vertex to: $(v,to)$\n",
    "\n",
    "Suppose that while performing DFS from $to$ vertex, there is no path of back edges to the vertex v or any of its ancestors.\n",
    "\n",
    "Then $(v,to)$ is a bridge, there is no other path from v to *to* except $(v,to)$ edge.\n",
    "\n",
    "Otherwise two edge-disjoint paths between v and to exist.\n",
    "\n",
    "![alternatvie text](img/dfs_bridges.png)\n",
    "\n",
    "To find a bridge we need to check that this condition is met.\n",
    "\n",
    "During DFS for each vertex v compute value $lowest[v]$:\n",
    "\n",
    "$$lowest[v]=\\begin{equation}\n",
    "min\\left\\{ \n",
    "  \\begin{aligned}\n",
    "    entry[v]\\\\\n",
    "    entry[p]\\\\\n",
    "    lowest[to]\\\\\n",
    "  \\end{aligned}\n",
    "  \\right.\n",
    "\\end{equation}\n",
    "$$\n",
    "\n",
    "where:\n",
    "- entry - entry time in DFS\n",
    "- $p$ is a gray vertex from back edge $(v,p)$, \n",
    "- $to$ is a child vertix of v in spanning tree of DFS\n",
    "\n",
    "$(v,to)$ is a bridge if when leaving vertex to\n",
    "\n",
    "$lowest[to] > entry[v]$\n",
    "\n",
    "We failed to find any child vertex of $to$ that have back edge connected with the vertex which entry time is less than $entry[v]$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6770e36e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 0 1 3 4 5 Bridge found (0,3)\n"
     ]
    }
   ],
   "source": [
    "class Graph:\n",
    "    def __init__(self):\n",
    "         self.graph = {}\n",
    "    # function to add an edge to graph\n",
    "    def addEdge(self, u, v):\n",
    "        if u not in self.graph:\n",
    "            self.graph[u] = []\n",
    "        self.graph[u].append(v)\n",
    " \n",
    "    def dfs(self, v, visited, entry, lowest, time, p = -1):\n",
    "        visited.add(v)\n",
    "        time = time+1\n",
    "        entry[v] = time\n",
    "        lowest[v] = time\n",
    "        print(v, end=' ')\n",
    "        for to in self.graph[v]:\n",
    "            if to == p:\n",
    "                continue\n",
    "            if to not in visited:\n",
    "                self.dfs(to, visited, entry, lowest, time, v)\n",
    "                lowest[v] = min(lowest[v], lowest[to])\n",
    "                if lowest[to] > entry[v]:\n",
    "                    print(f\"Bridge found ({v},{to})\")\n",
    "            else:\n",
    "                lowest[v] = min(lowest[v], entry[to])\n",
    " \n",
    "    def DFSMain(self, v):\n",
    "        visited = set()\n",
    "        entry = {}\n",
    "        lowest = {}\n",
    "        time = 0\n",
    "        self.dfs(v, visited, entry, lowest, time)\n",
    "        \n",
    "g = Graph()\n",
    "g.addEdge(0, 1)\n",
    "g.addEdge(1, 2)\n",
    "g.addEdge(2, 0)\n",
    "g.addEdge(3, 4)\n",
    "g.addEdge(4, 5)\n",
    "g.addEdge(5, 3)\n",
    "g.addEdge(0, 3)\n",
    "g.DFSMain(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a9c8df2",
   "metadata": {},
   "source": [
    "## Vertex-connection\n",
    "\n",
    "Two edges in a graph are **vertex-connected** if two vertex-disjoint paths that connect their ends exist.\n",
    "\n",
    "![alternatvie text](img/vertex_connection.png)\n",
    "\n",
    "All graph edges are partitioned into vertex-connected equivalence classes:\n",
    "\n",
    "- $(a,b) \\sim (a,b)$\n",
    "- $(a,b) \\sim (c,d) \\implies (c,d) \\sim (a,b)$\n",
    "- $(a,b) \\sim (c,d), (c,d) \\sim (e,f) \\implies (a,b) \\sim (e,f)$\n",
    "\n",
    "Let's show transitivity:\n",
    "\n",
    "- A path from f to d and a path from e to c form a cycle $C$\n",
    "- P is the path from a to c and Q is the path from b to d, P and Q intersect the cycle C in vertices x and y\n",
    "\n",
    "We can find two vertex-disjoint paths that connect b and e and a and f:\n",
    "\n",
    "- Path1: b -> y -> e\n",
    "- Path2: a -> x -> c -> d -> f\n",
    "\n",
    "![alternatvie text](img/vert_conn_trans.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee30e06d",
   "metadata": {},
   "source": [
    "All graph edges are partitioned into vertex-connected equivalence classes:\n",
    "\n",
    "![alternatvie text](img/art_points.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ecfb998",
   "metadata": {},
   "source": [
    "**Articulation point** is \n",
    "\n",
    "1) a vertex, that is incident to edges belonging to two or more vertex-connected connected components\n",
    "\n",
    "2) a vertex, upon removal of which, together with the edges incident to it, the connected component breaks up\n",
    "\n",
    "$1 \\implies 2$ : If we look at any pair of vertices connected to an articulation point with edges from different vertex-connected connected components, a path that connects these two vertices but does not contain the articulation point does not exist. If we remove the articulation point, the connected component breaks up\n",
    "\n",
    "$2 \\implies 1$ : Suppose that incident edges of a removed articulation point belong to one vertex-connected connected component. Then any two neighbors of the articulation point are from the edges of the same vertex-connected component. Then a path exists that connects this two vertices but does not contain the articulation point. It means that any two vertices connected to the articulation point stay in one connected component."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20653ce2",
   "metadata": {},
   "source": [
    "### Finding articulation points in graph\n",
    "\n",
    "We can find articulation points in graph using DFS: \n",
    "\n",
    "1) Check all vertices except the starting one\n",
    "\n",
    "2) Check starting vertex\n",
    "\n",
    "- Let $v$ not be the starting vertex. Then if there is a child of vertex  $v$ in the DFS tree that does not have a back edge in any of $v$ ancestors, then this vertex is an articulation point\n",
    "- If $v$ is the starting vertex, then it is an articulation point if and only if unvisited vertices exist among those connected to $v$ after the DFS from the first vertex connected to $v$.\n",
    "\n",
    "\n",
    "$$lowest[v]=\\begin{equation}\n",
    "min\\left\\{ \n",
    "  \\begin{aligned}\n",
    "    entry[v]\\\\\n",
    "    entry[p]\\\\\n",
    "    lowest[to]\\\\\n",
    "  \\end{aligned}\n",
    "  \\right.\n",
    "\\end{equation}\n",
    "$$\n",
    "\n",
    "$(v,to)$ is a bridge if when leaving vertex to\n",
    "\n",
    "$v$ is not the starting vertex, $v$ is an articulation point if vertex $to$ exist such as when leaving vertex $to$:\n",
    "\n",
    "$lowest[to] \\geq entry[v]$\n",
    "\n",
    "If $v$ is the starting vertex, then it is an articulation point if, after processing one adjacent vertex, the graph still has unprocessed vertices connected to $v$. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7ee2c9d",
   "metadata": {},
   "source": [
    "# Shortest path problem\n",
    "\n",
    "- Single pair shortest path problem - finding the shortest path between two vertices:\n",
    "\n",
    "Dijkstra's algorithm, A*, IdaStar\n",
    "- Single source shortest paths problem - finding the shortest paths from the selected vertex to all the others:\n",
    "\n",
    "Dijkstra's algorithm, Bellman–Ford algorithm\n",
    "- All pairs shortest paths problem - finding shortest paths between all pairs of vertices:\n",
    "\n",
    "Floyd–Warshall algorithm, Johnson's algorithm\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e875967",
   "metadata": {},
   "source": [
    "## Dijkstra's algorithm\n",
    "\n",
    "$G$ - directed weighted graph, all weights are non-negative. $s$ is a starting vertex.\n",
    "\n",
    "Dijkstra's algorithm works like BFS, but uses a priority queue instead of a regular queue\n",
    "\n",
    "If we reach a vertex that is already in the queue using shorter path, then this vertex will change its position in the priority queue.\n",
    "\n",
    "The next vertex extracted from the queue is the unvisited vertex closest to the start\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e7c887b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relaxation\n",
    "# d - shortest path weight, w -weights, pi - parent from the shortest path, u, v - vertices\n",
    "def relax(d, w, pi, u, v):\n",
    "    if d[v] > d[u] + w[(u,v)]:\n",
    "        d[v] = d[u] + w[(u,v)]\n",
    "        pi[v] = u"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f3c5c3",
   "metadata": {},
   "source": [
    "![alternatvie text](https://upload.wikimedia.org/wikipedia/commons/5/57/Dijkstra_Animation.gif)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69d51a66",
   "metadata": {},
   "source": [
    "Look at the implementation at https://github.com/denizetkar/priority-queue/blob/main/graph.py\n",
    "\n",
    "T = adding node to queue + update node in priority queue\n",
    "\n",
    "$T = O(E*log(V)+V*log(V)) = O((E+V)*log(V))$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a056eb3b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
