{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bfa81203",
   "metadata": {},
   "source": [
    "# Sequence homology\n",
    "\n",
    "**Sequence homology** is the biological homology between DNA, RNA, or protein sequences, defined in terms of shared ancestry in the evolutionary history of life. Homology among DNA, RNA, or proteins is typically inferred from their nucleotide or amino acid sequence similarity. **Significant similarity** is strong evidence that two sequences are related by evolutionary changes from a common ancestral sequence. Alignments of multiple sequences are used to indicate which regions of each sequence are homologous. [(Wikipedia)](https://en.wikipedia.org/wiki/Sequence_homology)\n",
    "\n",
    "![Alignment](https://upload.wikimedia.org/wikipedia/commons/0/03/Alignment-Comparison-En.png)\n",
    "\n",
    "\n",
    "\n",
    "## Global alignment\n",
    "\n",
    "A global alignment aligns two sequences from beginning to end, aligning each letter in each sequence only once.\n",
    "\n",
    "### Needleman–Wunsch algorithm\n",
    "\n",
    "We assign a total score to each possible alignment. The calculation of the score is based on the scores for each pair of nucleotides in their current position.\n",
    "\n",
    "- Match: The two nucleotides at the current index are the same.\n",
    "\n",
    "- Mismatch: The two nucleotides at the current index are different.\n",
    "\n",
    "- Indel (Insertion or Deletion): The best alignment involves one nucleotide aligning to a gap in the other string.\n",
    "\n",
    "Simple scoring system:\n",
    "- Match: +1\n",
    "- Mismatch or Indel: −1\n",
    "\n",
    "#### Example:\n",
    "\n",
    "**GCATGCG**\n",
    "\n",
    "**GATTACA**\n",
    "\n",
    "**Building**\n",
    "\n",
    "1) Fill the first row: we move horizontally, which means that an insertion occurred in the first sequence. We add a penalty for indel to the previous value.\n",
    "\n",
    "2) Fill the first column: we move vertically, which means that an insertion occurred in the second sequence. We add a penalty for indel to the previous value.\n",
    "\n",
    "3) Fill the rest of the dynamic programming array. We move from the top-left cell to the bottom-right cell:\n",
    "\n",
    "![Alignment](img/needleman_build.png)\n",
    "\n",
    "**Alignment**\n",
    "\n",
    "To get the best alignment we move from the bottom-right cell to the top-left:\n",
    "\n",
    "\n",
    "![Alignment](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3f/Needleman-Wunsch_pairwise_sequence_alignment.png/440px-Needleman-Wunsch_pairwise_sequence_alignment.png)\n",
    "\n",
    "\n",
    "### Result\n",
    "\n",
    "**GCATG-CG**\n",
    "\n",
    "**G-ATTACA**\n",
    "\n",
    "Result alignment score = 0\n",
    "\n",
    "Time complexity: for two sequences of length $m$ and $n$: $O(mn)$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07480ebf",
   "metadata": {},
   "source": [
    "# Local alignment\n",
    "\n",
    "**Local alignments** are more useful for dissimilar sequences that are suspected to contain regions of similarity or similar sequence motifs within their larger sequence context.\n",
    "\n",
    "## Smith–Waterman algorithm\n",
    "\n",
    "The Smith–Waterman algorithm is a general local alignment method based on the same dynamic programming scheme but with additional choices to start and end at any place.\n",
    "\n",
    "One dp matrix can have a number of local alignments, in own example we expect it to contain 2 local alignments\n",
    "\n",
    "**Building**\n",
    "\n",
    "1)\tFirst row and first column are set to 0\n",
    "\n",
    "2) Fill the rest of the dynamic programming array similar to Needleman–Wunsch. However, if the result score is negative, score is set to 0\t\n",
    "\n",
    "**Alignment**\n",
    "\n",
    "Begin with the highest score, end when 0 is encountered\t\n",
    "\n",
    "![Alignment](https://sistemanalize.files.wordpress.com/2017/12/smith-waterman5.png)\n",
    "\n",
    "\n",
    "**HEAGAWGHEE**\n",
    "\n",
    "**PAWHEAE**\n",
    "\n",
    "...\n",
    "\n",
    "Results: \n",
    "\n",
    "Starting from 28:\n",
    "\n",
    "**AWGHE**\n",
    "\n",
    "**AW-HE**\n",
    "\n",
    "From 27:\n",
    "\n",
    "**AWGHEE**\n",
    "\n",
    "**AW-HEA**\n",
    "\n",
    "From 26:\n",
    "\n",
    "**AWGHE-E**\n",
    "\n",
    "**AW-HEAE**\n",
    "\n",
    "From 21:\n",
    "\n",
    "**HEA**\n",
    "\n",
    "**HEA**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "413a0cb6",
   "metadata": {},
   "source": [
    "# Longest common subsequence\n",
    "\n",
    "**T**A**G**TC**A**GT**ACA**GCT\n",
    "\n",
    "**T**C**G**AA**A**CC**ACA**AAA\n",
    "\n",
    "Result:\n",
    "\n",
    "**TGAACA**\n",
    "\n",
    "Scoring system:\n",
    "- Match: +1\n",
    "- Mismatch or Indel: 0\n",
    "\n",
    "Same solution for global and local alignment problem\n",
    "\n",
    "# Longest common substring\n",
    "\n",
    "AAAA**GCGC**AAAA\n",
    "\n",
    "TTTTTTTTTT**GCGC**\n",
    "\n",
    "Result:\n",
    "\n",
    "**GCGC**\n",
    "\n",
    "Scoring system:\n",
    "- Match: +1\n",
    "- Mismatch or Indel: $\\infty$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c14949",
   "metadata": {},
   "source": [
    "# Substitution matrix\n",
    "\n",
    "**A substitution matrix** describes the frequency at which a character in a nucleotide sequence or a protein sequence changes to other character states over evolutionary time. We can use such matrices to penalize differently the mismatches of two amino acids during alignment.\n",
    "\n",
    "Substituting an amino acid with another with similar properties is more likely to have a smaller impact on the structure and function of a protein than replacement with an amino acid with different properties.\n",
    "\n",
    "The matrix should reflect the extent to which the replacement of one amino acid with another is fixed in evolution. The matrix shows how similar amino acids are to each other.\n",
    "\n",
    "$M(a,b)$\n",
    "\n",
    "For such matrix alignment score \n",
    "\n",
    "$W_{ij} = max(W_{i-1,j-1} + M(S_i,S_j), W_{i-1,j} + indel, W_{i,j-1} + indel)$\n",
    "\n",
    "## How can we calculate M(a,b)?\n",
    "\n",
    "Let's look at two cases\n",
    "\n",
    "1) Two sequences are aligned (dependent), for example\n",
    "\n",
    "    X: x1, x2, x3, x4...\n",
    "\n",
    "      \n",
    "\n",
    "    Y: y1, y2, y3, y4...\n",
    "\n",
    "The likelihood of such an alignment:\n",
    "\n",
    "$P(X,Y|1) = P(x_1 \\cdot y_1)*P(x_2 \\cdot y_2)*P(x_3 \\cdot y_3)*...$\n",
    "\n",
    "2) Two sequences are not aligned  (independent)\n",
    "\n",
    "The likelihood:\n",
    "\n",
    "$P(X,Y|2) = P(x_1) * P(y_1) * P(x_2) * P(y_2) * P(x_3) * P(y_3) *...$\n",
    "\n",
    "Log likelihood ratio:\n",
    "\n",
    "$\\sum_i(log(\\frac{ P(x_i \\cdot y_i)}{P(x_i) * P(y_i)}))$\n",
    "\n",
    "$M_{ij} = log(\\frac{ P(x_i \\cdot y_i)}{P(x_i) * P(y_i)})$\n",
    "\n",
    "$P(x_i)$ and $P(y_i)$ - frequency of occurrence of amino acids in nature\n",
    "\n",
    "How to compute  $P(xi \\cdot yi)$? We need a reference database of true alignments from where we obtain the frequencies for each pair of aminoacids. How to obtain true alignments?\n",
    "\n",
    "1) **PAM** (point accepted mutation) matrix. Based on global alignments of closely related proteins. **PAM1** is the matrix calculated from comparisons of sequences with no more than 1% divergence but corresponds to 99% sequence identity.\t\n",
    "\n",
    "$PAM1 = log(\\frac{ P(x_i \\cdot y_i)}{P(x_i) * P(y_i)}) = log(\\frac{ P(x_i | y_i)P (y_i)}{P(x_i) * P(y_i)})$\n",
    "\n",
    "The Markov chain model of protein mutation:\n",
    "\n",
    "$P_2(x_i | z_i) = \\sum_{y_i}(P_1(x_i | y_i) \\cdot P_1(y_i | z_i)) = P_1^2(x_i | y_i)$\n",
    "\n",
    "$PAM2 = PAM1^2$\n",
    "\n",
    "$PAMN = PAM1^N$\n",
    "\n",
    "2) **BLOCKS** - database of local multiple alignments without indels\n",
    "\n",
    "3) **BLOSUM** (BLOcks SUbstitution Matrix) \n",
    "\n",
    "BLOCKS database was scanned for very conserved regions of protein families and then the relative frequencies of amino acids and their substitution probabilities were counted.\n",
    " \n",
    "There can be many almost identical sequences in a block. We can use a threshold for the percentage of identity. Thus, the block must contain sequences that have no less than a specified number of mismatches. Throwing out similar sequences eliminates bias in the database.\n",
    "\n",
    "**BLOSUM80**: more related proteins\n",
    "\n",
    "**BLOSUM62**: midrange\n",
    "\n",
    "**BLOSUM45**: distantly related proteins\n",
    "\n",
    "\n",
    "\n",
    "BLOSUM looks directly at mutations in motifs of related sequences while PAM's extrapolate evolutionary information based on closely related sequences\n",
    "\n",
    "BLOSUM62:\n",
    "![x](https://upload.wikimedia.org/wikipedia/commons/thumb/f/f5/Blosum62-dayhoff-ordering.svg/1280px-Blosum62-dayhoff-ordering.svg.png)\n",
    "\n",
    "\n",
    "Larger numbers in BLOSUM matrix denote higher sequence similarity and therefore smaller evolutionary distance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "578dab69",
   "metadata": {},
   "source": [
    "# Multiple sequence alignment\n",
    "\n",
    "Alignment of 3 or more biological sequences\n",
    "\n",
    "**ClustalW** use progressive alignment approach\n",
    "\n",
    "Other multiple alignment algorithms: **MUSCLE**, **T-Coffee**, Prob Cons\n",
    "    \n",
    "Using multiple alignments of related sequences, we can find parts of the sequence that remain conserved through evolution\n",
    "\n",
    "### What problems can multiple alignment solve?\n",
    "\n",
    "- search for motifs and functional domains of proteins\n",
    "\n",
    "- construction of phylogenetic trees\n",
    "\n",
    "![x](https://upload.wikimedia.org/wikipedia/commons/7/79/RPLP0_90_ClustalW_aln.gif)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a863f1ae",
   "metadata": {},
   "source": [
    "## What if we use dynamic programming?\n",
    "\n",
    "For 3 sequences to align, number of transitions $2^3 - 1$\n",
    "\n",
    "![Title](img/multiple_align_dynamic.png)\n",
    "\n",
    "For $n$ sequences to align, number of transitions $2^n - 1$\n",
    "\n",
    "For sequences of length $L$: $(L^n)*(2^n - 1)$ ~ $(2L)^n$\n",
    "\n",
    "## Progressive alignment\n",
    "\n",
    "**Progressive alignment** is a heuristic technique that builds up a final alignment by combining pairwise alignments beginning with the most similar pair and progressing to the most distantly related. \n",
    "\n",
    "![Title](https://www.researchgate.net/publication/332443283/figure/fig2/AS:11431281179435865@1691186462008/ClustalW-based-progressive-alignment.png)\n",
    "\n",
    "1) Build similarity matrix based on pairwise distances between sequences\n",
    "\n",
    "2) Build guide tree, for example, based on cluster analysis using UPGMA\n",
    "\n",
    "3) Align sequences sequentially according to the guide tree, align neigboor nodes first\n",
    "\n",
    "![Title](https://www.researchgate.net/profile/Ken-Nguyen-4/publication/221865314/figure/fig1/AS:213385825263617@1427886553160/A-progressive-multiple-sequence-alignment-An-example-of-progressive-Multiple-Sequence.png)\n",
    "\n",
    "## How to align block of sequences?\n",
    "\n",
    "To align two blocks of sequences using dynamic programming we need to compute alignment score "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76f710c2",
   "metadata": {},
   "source": [
    "\n",
    "For each column in alignment of block X and block Y alignment score is a sum of mathc scores for all pairs:\n",
    "\n",
    "$w = \\sum_{(i,j)}(M(X_i, Y_j)+M(X_i, X_i) + M(Y_j,Y_j))$\n",
    "\n",
    "Example:\n",
    "For column\n",
    "\n",
    "    X: A\n",
    "\n",
    "       G\n",
    "\n",
    "       A\n",
    "\n",
    "    Y: -\n",
    "   \n",
    "       V\n",
    "\n",
    "$w = M(A,-) + M(A,V) + M(G,-) + M(G,V) + M(A,-) + M(-,V) + M(A,G) + M(G,A) + M(A,A) + M(-,A)$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b119a4d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A' 'T' 'T']\n",
      "['G' 'C' 'C']\n",
      "['C' 'A' 'A']\n",
      "['A' 'T' '-']\n",
      "['C' 'C' 'C']\n",
      "['T' 'T' 'T']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([['G', 'G', 'G'],\n",
       "       ['A', 'T', 'T'],\n",
       "       ['G', 'C', 'C'],\n",
       "       ['C', 'A', 'A'],\n",
       "       ['A', 'T', '-'],\n",
       "       ['C', 'C', 'C'],\n",
       "       ['T', 'T', 'T']], dtype='<U1')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "class AlignmentBlock:\n",
    "    def __init__(self, aligns):\n",
    "        self.block = np.array([list(x) for x in aligns]).T\n",
    "                \n",
    "    def __iter__(self):\n",
    "        return self.block.__iter__()\n",
    "    \n",
    "    def __getitem__(self, i):\n",
    "            return self.block[i]\n",
    "    def __len__(self):\n",
    "        return self.block.shape[0]\n",
    "    def concat(self, column):\n",
    "        if self.block.size == 0:\n",
    "            self.block = np.hstack((column, self.block))\n",
    "        else:\n",
    "            self.block = np.vstack((column, self.block))\n",
    "        \n",
    "    def num(self):\n",
    "        return self.block.shape[1]\n",
    "    def __str__(self):\n",
    "        return '\\n'.join([''.join(s) for s in self.block.T])\n",
    "    def join(self, block2):\n",
    "        self.block = np.hstack((self.block, block2.block))\n",
    "        \n",
    "a = AlignmentBlock([\"AGCACT\", \"TCATCT\", \"TCA-CT\"])\n",
    "b= AlignmentBlock([\"AGCACT\", \"TCATCT\", \"TCA-CT\"])\n",
    "\n",
    "#a.join(b)\n",
    "for i in a:\n",
    "    print(i)\n",
    "\n",
    "\n",
    "column = np.array(['G','G','G'])\n",
    "a.concat(column)\n",
    "a.block\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "504f57a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def msa_score(seq_block_column1, seq_block_column2):\n",
    "    result_score = 0\n",
    "    for a in seq_block_column1:\n",
    "        for b in seq_block_column2:\n",
    "            if a!=b:\n",
    "                result_score = result_score - 1 # -1 or result_score - blosum_matrix[a,b]\n",
    "            else:\n",
    "                result_score = result_score + 1 # match\n",
    "    return result_score\n",
    "\n",
    "msa_score(['a','b', 'b'], ['b', 'c'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a4fda2c",
   "metadata": {},
   "source": [
    "(a,b) -1, (b,b) 1, (b,b) 1, (a,c) -1 (b,c) -1 (b,c) -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ed527d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def smith_waterman_MSA(seq_block1, seq_block2, match_score=1, mismatch_score=-1, gap_penalty=-1):\n",
    "    score_matrix = np.zeros((len(seq_block1) + 1, len(seq_block2) + 1))\n",
    "    traceback_matrix = np.zeros((len(seq_block1) + 1, len(seq_block2) + 1), dtype=str)\n",
    "\n",
    "    for i in range(1, len(seq_block1) + 1):\n",
    "        for j in range(1, len(seq_block2) + 1):\n",
    "            match = score_matrix[i - 1][j - 1] + msa_score(seq_block1[i - 1], seq_block2[j - 1])\n",
    "            delete = score_matrix[i - 1][j] + gap_penalty\n",
    "            insert = score_matrix[i][j - 1] + gap_penalty\n",
    "            max_score= max(0, match, delete, insert)\n",
    "            score_matrix[i][j] = max_score\n",
    "\n",
    "            if max_score == match:\n",
    "                traceback_matrix[i][j] = \"M\"\n",
    "            elif max_score == delete:\n",
    "                traceback_matrix[i][j] = \"D\"\n",
    "            elif max_score == insert:\n",
    "                traceback_matrix[i][j] = \"I\"\n",
    "            else:\n",
    "                traceback_matrix[i][j] = \"E\"  # E represents the end of the alignment\n",
    "    # Find the maximum score and its indices in the score matrix\n",
    "    max_score = np.max(score_matrix)\n",
    "    max_i, max_j = np.unravel_index(np.argmax(score_matrix), score_matrix.shape)\n",
    "    \n",
    "    # Perform the traceback to retrieve the aligned sequences\n",
    "    aligned_block1 = AlignmentBlock([])\n",
    "    aligned_block2 = AlignmentBlock([])\n",
    "    i, j = max_i, max_j\n",
    "    while i > 0 and j > 0 and score_matrix[i][j] > 0:\n",
    "        if traceback_matrix[i][j] == \"M\":\n",
    "            aligned_block1.concat(seq_block1[i-1])\n",
    "            aligned_block2.concat(seq_block2[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif traceback_matrix[i][j] == \"D\":\n",
    "            aligned_block1.concat(seq_block1[i-1])\n",
    "            aligned_block2.concat([\"-\"] * seq_block2.num())\n",
    "            i -= 1\n",
    "        elif traceback_matrix[i][j] == \"I\":\n",
    "            aligned_block1.concat([\"-\"] * seq_block1.num())\n",
    "            aligned_block2.concat(seq_block2[j-1])\n",
    "            j -= 1\n",
    "    aligned_block1.join(aligned_block2)\n",
    "    return aligned_block1, max_score        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55c5be9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEDLQVRDVELAGAPGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN\n"
     ]
    }
   ],
   "source": [
    "insulin_sequences = [\n",
    "'MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN',\n",
    "'MALWMRFLPLLALLFLWESHPTQAFVKQHLCGSHLVEALYLVCGERGFFYTPMSRREVEDPQVAQLELGGGPGAGDLQTLALEVAQQKRGIVDQCCTSICSLYQLENYCN',\n",
    "'MALWTRLLALLALLALGAPTPARAFANQHLCGSHLVEALYLVCGERGFFYTPKARREVEDTQVGGVELGGGPGAGGLQPLGPEGRPQKRGIVEQCCASVCSLYQLENYCN',\n",
    "'MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEDLQVRDVELAGAPGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN']\n",
    "\n",
    "seq1 = AlignmentBlock([insulin_sequences[0]])\n",
    "seq2 = AlignmentBlock([insulin_sequences[3]])\n",
    "\n",
    "result,c = smith_waterman_MSA(seq1,seq2, match_score=1, mismatch_score=-1, gap_penalty=-1)\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a1c6bf33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MALWTRLLALLALLALGAPTPARAFANQHLCGSHLVEALYLVCGERGFFYTPKARREVEDTQVGGVELGGGPGAGGLQPLGPEGRPQKRGIVEQCCASVCSLYQLENYCN\n",
      "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
      "MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEDLQVRDVELAGAPGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN\n"
     ]
    }
   ],
   "source": [
    "seq3 = AlignmentBlock([insulin_sequences[2]])\n",
    "result2,c =smith_waterman_MSA(seq3, result, match_score=1, mismatch_score=-1, gap_penalty=-1)\n",
    "print(result2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "24d70146",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MALWMR-FLPLLALL-FLW---ESHPT-QAFV-KQHLCGSHLVEALYLVCGERGFFYTP--MSRREVED-PQV-AQ-LELGGGPGAG-DLQ-TLALE-VA-QQKRGIV-DQCCTSICSLYQLENYCN\n",
      "MALWTRL-LALLALLA-LGAPT---PAR-AFAN-QHLCGSHLVEALYLVCGERGFFYTPKA--RREVEDT-QVG-GV-ELGGGPGAGG-LQP-LGPEG-RP-QKRGIVE-QCCASVCSLYQLENYCN\n",
      "MALWMRL-LPLLALLA-LWGPD---PAA-AFVN-QHLCGSHLVEALYLVCGERGFFYTPKT--RREAEDL-QVG-QV-ELGGGPGAGS-LQP-LALEG-SL-QKRGIVE-QCCTSICSLYQLENYCN\n",
      "MALWMRL-LPLLALLA-LWAPA---PTR-AFVN-QHLCGSHLVEALYLVCGERGFFYTPKA--RREVEDL-QVR-DV-ELAGAPGEGG-LQP-LALEG-AL-QKRGIVE-QCCTSICSLYQLENYCN\n"
     ]
    }
   ],
   "source": [
    "seq4 = AlignmentBlock([insulin_sequences[1]])\n",
    "result3,c =smith_waterman_MSA(seq4, result2, match_score=1, mismatch_score=-1, gap_penalty=-1)\n",
    "print(result3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "993cfb96",
   "metadata": {},
   "source": [
    "# UPGMA (unweighted pair group method with arithmetic mean)\n",
    "\n",
    "UPGMA is an agglomerative (bottom-up) hierarchical clustering method.  At each step, the most similar two clusters are merged into a higher-level cluster. \n",
    "\n",
    "Input: D - similarity matrix\n",
    "\n",
    "Output: dendrogram (guide tree)\n",
    "\n",
    "UPGMA defines the distance between clusters C1 and C2 as the average pairwise distance between elements of C1 and C2:\n",
    "\n",
    "$D_{C_1,C_2} = \\frac{\\sum_{i \\in C}\\sum_{i \\in C_2}D_{i,j}}{|C_1|\\cdot|C_2|}$\n",
    "\n",
    "When $C_1$ and $C_2$ are merged into C, node C become a parent in the guide tree.\n",
    "\n",
    "UPGMA build ultrametric guide tree. **An ultrametric tree** is a rooted tree with edge lengths where all leaves are equidistant from the root. The height of each node is set to $\\frac{D_{C_i, C_j}}{2}$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e14cde42",
   "metadata": {},
   "source": [
    "![Title](http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma15.PNG)\n",
    "\n",
    "http://www.slimsuite.unsw.edu.au/teaching/upgma/\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
